File: BaekJoon/Graph/BFS/치즈_2636.cpp
//
//  치즈_2636.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/02.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: 치즈 2636

// MARK: - BFS 문제풀이
#if 0
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
typedef pair<int,int> Pair;

int tx[] = {0,0,-1,1};
int ty[] = {-1,1,0,0};
int Cnt = 0, Time = 1;

vector<vector<int>> V(101, vector<int>(101,0));
vector<vector<int>> C(101, vector<int>(101,0));
vector<Pair> rV;

void BFS(int X, int Y, const int &M, const int &N) {
    queue<Pair> Q;
    C[X][Y]=0;
    Q.push({X,Y});
    while(!Q.empty()) {
        auto node = Q.front();
        Q.pop();
        int nx = node.first;
        int ny = node.second;
        for(int i=0; i<4; i++) {
            int mx = nx + tx[i];
            int my = ny + ty[i];
            if(mx<0 || my<0 || mx>=M || my>=N) continue;
            if(C[mx][my]==1) continue;
            C[mx][my]=1;
            if(V[mx][my]==1) rV.push_back({mx,my});
            else Q.push({mx,my});
        }
    }
}

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0);
    int M,N; cin>>M>>N;
    for(int i=0; i<M; i++)
        for(int j=0; j<N; j++) {
            cin>>V[i][j];
            Cnt += V[i][j]==1 ? 1 : 0;
        }
    
    if(Cnt==0) return !printf("%d\n%d\n",0,0);
    
    while(Cnt>0) {
        C = vector<vector<int>>(101, vector<int>(101,0));
        rV = vector<Pair>();
        int nowCnt = Cnt;
        BFS(0,0,M,N);
        if(Cnt-rV.size()==0) return !printf("%d\n%d\n",Time,nowCnt);
        for(auto &v: rV) V[v.first][v.second]=0;
        Cnt -= rV.size();
        Time++;
    }
    return 0;
}
#endif

File: BaekJoonPS/BaekjoonAlgorithmPS/Graph/_BFS/치즈_2636.swift
//
//  치즈_2636.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/03.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: 치즈 2636

// MARK: - BFS 문제풀이

/*
 import Foundation

 typealias Pair = (Int,Int)
 let arr = readLine()!.split(separator: " ").map { Int($0)! }
 let M = arr[0], N = arr[1]
 let tx = [-1,1,0,0]
 let ty = [0,0,-1,1]
 var G = [[Int]](repeating: [Int](), count: M)
 var tempG = [[Int]]()
 var C = [[Int]]()
 var Cnt = 0, Time = 1

 for i in G.indices {
 G[i] = readLine()!.split(separator: " ").map { Int($0)! }
 }

 for i in G.indices {
 for j in G[i].indices {
     if G[i][j]==1 { Cnt += 1 }
 }
 }

 tempG = G

 func BFS(_ X: Int, _ Y: Int) {
 var Q = [Pair]()
 Q.append((X,Y))
 while !Q.isEmpty {
     let node = Q.first!
     Q.removeFirst()
     let nx = node.0
     let ny = node.1
     for i in 0..<4 {
         let mx = nx + tx[i]
         let my = ny + ty[i]
         if mx<0 || my<0 || mx>=M || my>=N { continue }
         if C[mx][my]==1 { continue }
         C[mx][my]=1
         if G[mx][my]==1 {
             tempG[mx][my]=0
             Cnt -= 1
         }
         else {
             Q.append((mx,my))
         }
     }
 }
 }

 while Cnt > 0 {
 let nowCnt = Cnt
 C = [[Int]](repeating: [Int](repeating: 0, count: N), count: M)
 C[0][0]=1
 BFS(0,0)
 G = tempG
 if Cnt <= 0 {
     print(Time); print(nowCnt)
     break
 }
 Time += 1
 }
 */
